nifi-nar-bundles/nifi-mongodb-bundle/nifi-mongodb-services/src/main/java/org/apache/nifi/mongodb/MongoDBLookupService.java

```diff
@@ -52,68 +54,125 @@
-
-    static {
-        lookupDescriptors = new ArrayList<>();
-        lookupDescriptors.addAll(descriptors);
-        lookupDescriptors.add(LOOKUP_VALUE_FIELD);
-    }
-
     @Override
     public Optional<Object> lookup(Map<String, Object> coordinates) throws LookupFailureException {
-        Map<String, Object> clean = new HashMap<>();
-        clean.putAll(coordinates);
+        Map<String, Object> clean = coordinates.entrySet().stream()
+            .filter(e -> !schemaNameProperty.equals(String.format("${%s}", e.getKey())))
+            .collect(Collectors.toMap(
+                e -> e.getKey(),
+                e -> e.getValue()
+            ));
         Document query = new Document(clean);
 
         if (coordinates.size() == 0) {
             throw new LookupFailureException("No keys were configured. Mongo query would return random documents.");
         }
 
         try {
-            Document result = this.findOne(query);
+            Document result = projection != null ? controllerService.findOne(query, projection) : controllerService.findOne(query);
 
             if(result == null) {
                 return Optional.empty();
             } else if (!StringUtils.isEmpty(lookupValueField)) {
                 return Optional.ofNullable(result.get(lookupValueField));
             } else {
-                final List<RecordField> fields = new ArrayList<>();
+                RecordSchema schema = loadSchema(coordinates);
 
-                for (String key : result.keySet()) {
-                    if (key.equals("_id")) {
-                        continue;
-                    }
-                    fields.add(new RecordField(key, RecordFieldType.STRING.getDataType()));
-                }
-
-                final RecordSchema schema = new SimpleRecordSchema(fields);
-                return Optional.ofNullable(new MapRecord(schema, result));
+                RecordSchema toUse = schema != null ? schema : convertSchema(result);
+                return Optional.ofNullable(new MapRecord(toUse, result));
             }
         } catch (Exception ex) {
             getLogger().error("Error during lookup {}", new Object[]{ query.toJson() }, ex);
             throw new LookupFailureException(ex);
         }
     }
 
-    @Override
+    private RecordSchema loadSchema(Map<String, Object> coordinates) {
+        Map<String, String> variables = coordinates.entrySet().stream()
+            .collect(Collectors.toMap(
+                e -> e.getKey(),
+                e -> e.getValue().toString()
+            ));
+        try {
+            return getSchema(variables, null);
+        } catch (Exception ex) {
+            return null;
+        }
+    }
+
+    private RecordSchema convertSchema(Map<String, Object> result) {
+        List<RecordField> fields = new ArrayList<>();
+        for (Map.Entry<String, Object> entry : result.entrySet()) {
+
+            RecordField field;
+            if (entry.getValue() instanceof Integer) {
```

Comment:
At some point once this and the ES Lookup Service are merged, I'll refactor this into a helper method.